Sprint 4/step 1

Kvisaz opened the pull request:
Добрый день
Добавил новые страницы, 
переписал показ сообщения и ленту заказов на работу с сокетами (изначально ловил ленту заказов на ответ REST, потерял время)

Commit:
sprint-4/step-1 - fix e.toString

Commit:
sprint-4/step-1 - Routes and prettier

Commit:
sprint-4/step-1 - app.tsx prettier

Commit:
sprint-4/step-1 - draft components && routes

Commit:
sprint-4/step-1 - OrderDetails prettier

Commit:
sprint-4/step-1 - total prettier

Commit:
sprint-4/step-1 - orderSuccessResults, orderFeed in state

Commit:
sprint-4/step-1 - get orderFeed && order mock

Commit:
sprint-4/step-1 - live orderFeed

Commit:
sprint-4/step-1 - remove logging

Commit:
sprint-4/step-1 - formatDate

Commit:
sprint-4/step-1 - ordert status

Commit:
sprint-4/step-1 - OrderFeedItemStatus

Commit:
sprint-4/step-1 - order feed item

Commit:
sprint-4/step-1 - service refactoring

Commit:
sprint-4/step-1 - maxLength prop

Commit:
sprint-4/step-1 - OrderFeedPage

Commit:
sprint-4/step-1 - OrderFeedPage 2 cols

Commit:
sprint-4/step-1 - OrderFeedInfo

Commit:
sprint-4/step-1 - OrderFeed withStatus

Commit:
sprint-4/step-1 - routes

Commit:
sprint-4/step-1 - ProfileEditor extract

Commit:
sprint-4/step-1 - ProfileEditor routing

Commit:
sprint-4/step-1 - css tuning

Commit:
sprint-4/step-1 - css tuning 2

Commit:
sprint-4/step-1 - order details open

Commit:
sprint-4/step-1 - IRichOrderData

Commit:
sprint-4/step-1 - IOrderFeedItem naming

Commit:
sprint-4/step-1 - order feed item components

Commit:
sprint-4/step-1 - order feed details layout

Commit:
sprint-4/step-1 - OrderFeedDetailsPartList items

Commit:
sprint-4/step-1 - OrderFeedDetailsPartList overlapping

Commit:
sprint-4/step-1 - formatOrderNumber

Commit:
sprint-4/step-1 - order feed details margin right

Kvisaz committed:
sprint-4/step-1 - scrollable details list

Kvisaz committed:
sprint-4/step-1 - order with token

Kvisaz committed:
sprint-4/step-1 - orderSocketMiddleWare

Kvisaz committed:
sprint-4/step-1 - orderSocketMiddleWare update

Kvisaz committed:
sprint-4/step-1 - useMainState

Kvisaz committed:
sprint-4/step-1 - orderSocketMiddleWare refac

Kvisaz committed:
sprint-4/step-1 - logg and order

Kvisaz committed:
sprint-4/step-1 - OrderFeedInfo use real data

Kvisaz committed:
sprint-4/step-1 - logg off

Kvisaz committed:
sprint-4/step-1 - show order by ws event

gennady-bars commented on src/services/middleware/socketMiddleWare.ts:
```diff
@@ -0,0 +1,73 @@
+  onMessage: IBurgerActionType;
+}
+
+export const socketMiddleWare = (wsUrl: string, wsActions: IWSActions) => {
```
хорошо сделан `socketMiddleWare`

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -1,46 +1,46 @@
+    <div className={styles.content}>
+      <div className={`text text_type_main-medium ${styles.label}`}>Восстановление пароля</div>
+      <Input type={'email'} placeholder={'Укажите email'} value={email} onChange={onEmailChange} />
+      <Button onClick={onButtonClick} size={'medium'} type={'primary'}>Восстановить</Button>
```
` onClick={onButtonClick}`

Обработчик сабмита нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. Это нужно исправить везде, где есть инпуты

Обратите внимание, что кнопка сабмита должна быть внутри тега `form`, чтобы она запускала событие `submit`

gennady-bars commented on src/pages/forgot-password/forgot-password.tsx:
```diff
@@ -1,46 +1,46 @@
+  return (<div className={styles.wrap}>
+    <div className={styles.content}>
+      <div className={`text text_type_main-medium ${styles.label}`}>Восстановление пароля</div>
+      <Input type={'email'} placeholder={'Укажите email'} value={email} onChange={onEmailChange} />
```
Если есть инпуты, то они должны быть внутри тега `form` и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.

gennady-bars reviewed (commented):
Здравствуйте. Работа проделана огромная, но есть некоторые недочеты:

* если открыть детали заказа в ленте и перезагрузить  страницу, то на эту страницу не попадаю. Скрин http://joxi.ru/D2PnJ9QCBlb8pr . Но после авторизации получилось. А лента без авторизации должна работать
* при входе в ленту заказов в консоли ошибка `key`. Скрин http://joxi.ru/V2V749KC8MnzEA
* Вам нужно разделить редьюсер на несколько, иначе очень много кода в одном редьюсере. С экшенами тоже самое
* Обработчик сабмита нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. Это нужно исправить везде, где есть инпуты
* Если есть инпуты, то они должны быть внутри тега `form` и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита. Это нужно исправить везде, где есть инпуты
* Обратите внимание, что кнопка сабмита должна быть внутри тега `form`, чтобы она запускала событие `submit`
* Вам нужно использовать средства роутинга для определения, показывать ли попап с деталями, или показывать страницу. Вам должны были дать ссылку на документацию по этой логике. Вот она https://reactrouter.com/web/example/modal-gallery . В ней используются `location` и `background`, которые передаются в `Switch`. А еще для этого нужно проверять в `location` поле `action` и был ли там `PUSH` или `REPLACE`. Пожалуйста, посоветуйтесь с наставником.
* При входе на страницу ленты заказов нужно открывать соединение сокета, а при покидании страницы – закрывать. Это делается все в одном хуке `useEffect`. Тоже самое и с историей заказов . Нигде больше этих открытий не должно быть



**Можно лучше**

*  при клике на логотип обычно идет переход на главную страницу

Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта перед отправкой на ревью. И не забудьте, пожалуйста,  проверить наличие ошибок в консоли браузера перед отправкой (кнопка `F12`)

Удачного рефакторинга кода.

Kvisaz committed:
sprint-4/step-1 - extract ingredients reducer

Kvisaz committed:
sprint-4/step-1 - IngredientStorage

Kvisaz committed:
sprint-4/step-1 - ordersFeedReducer

Kvisaz committed:
sprint-4/step-1 - MainAction rename

Kvisaz committed:
sprint-4/step-1 - main actions folder

Kvisaz committed:
sprint-4/step-1 - ApiService extracted

Kvisaz committed:
sprint-4/step-1 - INGREDIENTS_ACTION extracted

Kvisaz committed:
sprint-4/step-1 - MAIN_ACTION extracted

Kvisaz committed:
sprint-4/step-1 - ORDERS_ACTION extracted

Kvisaz committed:
sprint-4/step-1 - updateOrderFeedFromHttp in App hook

Kvisaz committed:
sprint-4/step-1 - form submit

Kvisaz committed:
sprint-4/step-1 - form submit preventDefault

Kvisaz committed:
sprint-4/step-1 - useOrderFeed hook

Kvisaz committed:
sprint-4/step-1 - restoreAuth fix

Kvisaz committed:
sprint-4/step-1 - logo link

Kvisaz committed:
sprint-4/step-1 - useOrderFeed in order feed and order details

Kvisaz committed:
sprint-4/step-1 - fix logout

Kvisaz commented:
Спасибо за учебу.

Правки

1. FIXED: При перезагрузке теперь работают и детали заказа без авторизации (хотя потенциально это не гарантируется, так как без токена бэк шлет фейковый набор, как нам сообщили)

2. FIXED: Редьюсер, actions - разделил на три части, можно разделить еще, принцип понятен, просто нужно время. На работе уже используем слайсы из redux-toolkit, после того рефакторинга их польза особенно заметна. В следующий раз буду писать сразу со слайсами

3. FIXED: input помещены в form, обработчики отправки перемещены на submit формы 

4. APPROVED BEFORE:  По роутингу модальных окон - он уже был одобрен, мы его обсуждали на предыдущем этапе с Ириной Мозес, она это одобрила. Вкратце мои тезисы
- пример из react-router хорош, но он рассчитан на чистый лист, он не учитывает, что у нас уже есть redux
- этот параллельный стейт и логика - ухудшает код. Если есть redux, то лучше использовать его

5. FIXED: При входе на страницу ленты заказов нужно открывать соединение сокета, а при покидании страницы – закрывать. Это делается все в одном хуке useEffect. 

6. FIXED: при клике на логотип - переход на главную страницу

С уважением, Сергей Токарев.

Kvisaz commented:
@gennady-bars 
Ваш совет вынести сокеты в страницу был хорош и уместен - но я вспомнил, почему у меня было сделано именно так, как вы видели.

Дело в том, что бэк очень туго отдает ответ на авторизованный заказ, с большой паузой. Поэтому я изначально построил подключение сокетов при запуске приложения - и это отлично работало, потому что по сокетам можно было сразу задетектировать новый заказ - и приложение его показывало.

Сейчас хук сломал этот функционал, так как сокеты работают только на страницах с лентой.

Я сейчас подумаю, как это решить, сохранив ваш хук, но если бы мы работали в одной команде - я бы предложил вернуться к моему варианту. Особенно потому, что хук с сокетами не был описан в ТЗ, а функционал только ухудшился.

Kvisaz committed:
sprint-4/step-1 - fix order && order messages

Kvisaz committed:
sprint-4/step-1 - fix showCreatedOrder

Kvisaz commented:
@gennady-bars 
Последние фиксы - починил заказ, оставив сокеты там, где вы и попросили

Проблему долгого ожидания попапа решил через сообщения, которые появляются на месте конструктора.
Пока не придет success и fail - не сменится сообщение,  оно через 1.5 секунды уходит и открывает конструктор


![2021-10-12_004132](https://user-images.githubusercontent.com/16399450/136833560-901ce14e-ca8c-4494-af47-250582d6a253.png)
![2021-10-12_004141](https://user-images.githubusercontent.com/16399450/136833564-07f9418e-d67a-4b5b-a55a-9584ee8554dc.png)
![2021-10-12_004418](https://user-images.githubusercontent.com/16399450/136833565-cdf2c1df-88f6-4529-a630-cd4ad990625b.png)

gennady-bars commented:
> @gennady-bars Последние фиксы - починил заказ, оставив сокеты там, где вы и попросили
> 
> Проблему долгого ожидания попапа решил через сообщения, которые появляются на месте конструктора. Пока не придет success и fail - не сменится сообщение, оно через 1.5 секунды уходит и открывает конструктор
> 
> ![2021-10-12_004132](https://user-images.githubusercontent.com/16399450/136833560-901ce14e-ca8c-4494-af47-250582d6a253.png) ![2021-10-12_004141](https://user-images.githubusercontent.com/16399450/136833564-07f9418e-d67a-4b5b-a55a-9584ee8554dc.png) ![2021-10-12_004418](https://user-images.githubusercontent.com/16399450/136833565-cdf2c1df-88f6-4529-a630-cd4ad990625b.png)

очень достойно сделали правки. Супер!

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
